### ISSUE TITLE
JVM SIGSEGV after connection close with ingested arrow data.

### ISSUE THREAD

**Issue by jonathanswenson** [duckdb/duckdb#9528]
### What happens?

JVM crashed with a SIGSEGV after running query that was "cancelled". 
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
# SIGSEGV (0xb) at pc=0x00007fcef7bb6db5, pid=1, tid=226
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.9+9 (17.0.9+9) (build 17.0.9+9)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (17.0.9+9, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# C [libduckdb_java16057437192090020633.so+0x805db5] std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+0x65
# Core dump will be written. Default location: //core.1
# 
# An error report file with more information is saved as:
# /tmp/hs_err_pid1.log
# [error occurred during error reporting (), id 0xb, SIGSEGV (0xb) at pc=0x00007fcfb2b8e898]
#
# If you would like to submit a bug report, please visit:
# https://github.com/adoptium/adoptium-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

When this happened an Arrow Payload was loaded into duckdb using something like the following (kotlin)
```
ArrowStreamReader(inputStream, allocator).use { arrowReader ->
    ArrowArrayStream.allocateNew(allocator).use { arrowArrayStream ->
        Data.exportArrayStream(allocator, arrowReader, arrowArrayStream)
        duckDBConnection.registerArrowStream(tempTableName, arrowArrayStream)
        duckDBConnection.createStatement().use { statement ->
        statement.executeUpdate("Create TABLE \"$tableName\" AS SELECT * FROM \"${tempTableName}\"")
    }
}
```

Then, a query is executed against that same connection using the ingested table. 
At some point during execution, the query is killed, and attempting to extract the result (again using the arrow methods) fails with a `java.sql.SQLException: Result set is closed` error (likely expected). 

After the connection is closed, the above SIGSEGV kills the JVM. Every time we've seen, the container is also killed causing us to lose the core dump (if one is successfully written). 

### To Reproduce

I have not yet successfully managed to reproduce in a reliable way.

### OS:

Linux x86

### DuckDB Version:

0.9.1

### DuckDB Client:

java

### Full Name:

Jonathan Swenson

### Affiliation:

Omni

**Comment by jonathanswenson** [comment]
Potentially also related, I have noticed that upon ingesting data using the above method for arrowStreams, closing the connection and allocator often intermittently throws a IllegalStateException `RefCnt has gone negative`. 

If the resource isn't closed, closing the allocator results in an error indicating memory has been leaked. 

Perhaps there needs to be a way to unregister an arrow stream (manually) from the java side so that it can be freed before the connection is closed?

**Comment by jonathanswenson** [comment]
having trouble getting this to consistently reproduce, but managed to get the same behavior, but a slightly different error frame: 
```
# An error report file with more information is saved as:
# /tmp/hs_err_pid1.log
#
# A fatal error has been detected by the Java Runtime Environment:
#
# SIGSEGV (0xb) at pc=0x00007f9f4c9e659f, pid=1, tid=107
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.9+9 (17.0.9+9) (build 17.0.9+9)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.9+9 (17.0.9+9, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# v ~BufferBlob::jni_fast_GetLongField
#
# Core dump will be written. Default location: //core.1
#
```

**Comment by jonathanswenson** [comment]
My gut feel as though something about how I'm calling cancel is not allowed. 

Looking at the comment here:  https://github.com/duckdb/duckdb/blob/main/tools/jdbc/src/main/java/org/duckdb/DuckDBPreparedStatement.java#L339-L342

While I haven't managed to reproduce exactly the error I saw the first time around (at `[libduckdb_java16057437192090020633.so+0x805db5] std::_Sp_counted_base<(__gnu_cxx::_Lock_policy)2>::_M_release()+0x65`) I can reproduce the second error (at `~BufferBlob::jni_fast_GetLongField`) racing some threads. 

At a high level the cancellation is being called by a separate thread and while I haven't proven it, I can see it happening after the statement (at least) has been closed and potentially after the connection has been closed. 

From my first reproduction though, it apperars as though the cancel happens WHILE the connection is still open sometime before the arrowExport.

**Comment by pdet** [comment]
Hi @jonathanswenson 

I'm having a bit of trouble reproducing your error.

I've tried generating an Arrow Table over a parquet file, ingesting that, interrupting the ingestion from another thread, and trying to read the created table (And some variants of this workflow) but I did not manage to get this crash.

Could you provide a complete reproducible example?
Something similar to: https://github.com/pdet/reproducible_bug_example would be very helpful!

**Comment by jonathanswenson** [comment]
Here's a somewhat crude example, but it seems to fail for me every time. 

I'm using duckdb-jdbc 0.9.1 and arrow 12 (c data). 

I've played with a few different ways to play with the timing of the synchronize and the cancel, but this is probably the most simple that reproduces a few different ways. I can muck with the Thread.sleep(...) in there (0, 1, 2 ms) and get slightly different places where the SIGSEGV seems to originate. 

<details> <summary> java example that throws a SIGSEGV </summary>

```
import org.apache.arrow.c.ArrowArrayStream;
import org.apache.arrow.c.Data;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.IntVector;
import org.apache.arrow.vector.VarCharVector;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ArrowReader;
import org.apache.arrow.vector.ipc.ArrowStreamReader;
import org.apache.arrow.vector.ipc.ArrowStreamWriter;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.arrow.vector.types.pojo.FieldType;
import org.apache.arrow.vector.types.pojo.Schema;
import org.duckdb.DuckDBConnection;
import org.duckdb.DuckDBDriver;
import org.duckdb.DuckDBResultSet;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.*;
import java.util.Properties;

import static java.util.Arrays.asList;

public class DuckDBCancel {
    static int iterations = 1000;
    public static void main(String[] args) throws Exception{
        new DuckDBCancel().runTest();
    }

    public void runCancelTest(String filename) {
        Driver driver = new DuckDBDriver();
        String connectionString = "jdbc:duckdb:";
        for (int i = 0; i < iterations; i++) {
            try (BufferAllocator allocator = new RootAllocator()) {
                System.out.println("iteration " + i);
                try (Connection connection = driver.connect(connectionString, new Properties())) {

                    DuckDBConnection duckDBConnection = connection.unwrap(DuckDBConnection.class);
                    loadTable(filename, allocator, duckDBConnection);

                    try (Statement statement = connection.createStatement()) {
                        Thread t = new Thread(() -> {
                            try {
                                statement.cancel();
                                statement.close();
                            } catch (SQLException e) {
                                e.printStackTrace();
                            }
                        });

                        try (ResultSet resultSet = statement.executeQuery("SELECT * from foo")) {
                            t.start();
                            // just try to context switch
                            // depending on what number here I get SIGSEVs in different places.
                            Thread.sleep(0);

                            DuckDBResultSet duckDBResultSet = resultSet.unwrap(DuckDBResultSet.class);

                            try (ArrowReader reader = (ArrowReader) duckDBResultSet.arrowExportStream(allocator, 2048)) {
                                while (reader.loadNextBatch()) {
                                    // do nothing with the results
                                }
                            }
                        } finally {
                            t.join();
                        }
                    }
                } catch (Exception e) {
                    System.out.println("iteration " + i + " threw: " + e.getMessage());
                }
            }
        }
    }


    private void createArrowFile(File file) throws IOException {
        Field age = new Field("age", FieldType.nullable(new ArrowType.Int(32, true)),null);
        Field name = new Field("name", FieldType.nullable(new ArrowType.Utf8()), null);

        Schema schema = new Schema(asList(age, name), null);
        int rows = 100;
        try(BufferAllocator allocator = new RootAllocator();
            VectorSchemaRoot root = VectorSchemaRoot.create(schema, allocator);
            IntVector ageVector = (IntVector) root.getVector("age");
            VarCharVector nameVector = (VarCharVector) root.getVector("name")){

            root.setRowCount(rows);
            ageVector.allocateNew(rows);
            nameVector.allocateNew(rows);

            for (int i = 0; i < rows; i++) {
                ageVector.set(i, i);
                nameVector.set(i, ("Name " + i).getBytes(StandardCharsets.UTF_8));
            }

            try (OutputStream out = new FileOutputStream(file.getAbsolutePath());
                 ArrowStreamWriter writer = new ArrowStreamWriter(root, null, out)) {
                writer.writeBatch();
                writer.end();
            }
        }
    }

    public void runTest() throws Exception {
        File temp = File.createTempFile("duckdb", ".arrow");
        String filename = temp.getAbsolutePath();
        createArrowFile(temp);
        try {
            runCancelTest(filename);
        } catch (Exception e) {
            // of course the error here is loud enough that we never get to the finally...
            // nothing
        } finally {
            Files.delete(Path.of(filename));
        }
    }

    private void loadTable(String filename, BufferAllocator allocator, DuckDBConnection duckDBConnection) throws Exception {
        try (InputStream rawStream = new FileInputStream(filename);
             ArrowStreamReader streamReader = new ArrowStreamReader(rawStream, allocator);
             ArrowArrayStream arrowArrayStream = ArrowArrayStream.allocateNew(allocator)) {
            Data.exportArrayStream(allocator, streamReader, arrowArrayStream);
            duckDBConnection.registerArrowStream("tmp", arrowArrayStream);

            try (Statement stmt = duckDBConnection.createStatement()) {
                stmt.execute("CREATE TABLE foo as select * from tmp");
            }
        }
    }
}

```

</details>

This also reproduces the RefCnt going negative sometimes and usually crashes after a few iterations on my Apple Silicon Mac. I also tried it running on a linux aarch64 container and linux x86 container (emulated), but I don't see the negative RefCnt error.

**Comment by pdet** [comment]
Hey @jonathanswenson thanks for sending the code, I'm still having some trouble executing it in my machine, could you share the pom file with the exact dependencies and versions you used, and also share your java version?

**Comment by jonathanswenson** [comment]
I'm using gradle, but here's the contents of a sample gradle file / dependencies I'm using: 

```
plugins {
    java
}

group = "org.example"
version = "1.0-SNAPSHOT"

repositories {
    mavenCentral()
}

val arrowVersion = "12.0.0"

dependencies {
    implementation("org.apache.arrow:arrow-vector:$arrowVersion")
    implementation("org.apache.arrow:arrow-memory-unsafe:$arrowVersion")
    implementation("org.apache.arrow:arrow-jdbc:$arrowVersion")
    implementation("org.apache.arrow:arrow-c-data:$arrowVersion")
    implementation("org.duckdb:duckdb_jdbc:0.9.1")
}
```

I'm using JDK 17 (specifically open jdk Temurin - https://adoptium.net/temurin/releases/?version=17&os=mac&arch=aarch64) 

Locally I'm running on mac aarch64, but can also reproduce on a docker container based on the `eclipse-temurin:17` image with both the linux/amd64 and linux/aarch64 platform options. However the particular version of arrow that I have above does not work on aarch64 (but this has been fixed in arrow 14.0.0). While I have not tested 14.0.0 on those platforms, swapping to arrow 14.0.0 continues to reproduce the issue locally on my mac.

**Comment by pdet** [comment]
Alright, thanks for the extra info! I confirm I can reproduce it :-)!
